src/lib/expression/function_expression.cpp

```diff
@@ -0,0 +1,63 @@
+#include "function_expression.hpp"
+
+#include <sstream>
+
+#include "boost/functional/hash.hpp"
+
+#include "utils/assert.hpp"
+#include "constant_mappings.hpp"
+#include "expression_utils.hpp"
+
+namespace opossum {
+
+FunctionExpression::FunctionExpression(const FunctionType function_type,
+                                         const std::vector<std::shared_ptr<AbstractExpression>>& arguments):
+AbstractExpression(ExpressionType::Function, arguments), function_type(function_type) {
+
+  switch (function_type) {
+    case FunctionType::Substring:
+      Assert(arguments.size() == 3, "Substring expects 3 parameters");
+      break;
+    case FunctionType::Concatenate:
+      Assert(arguments.size() >= 2, "Concatenate expects at least 2 parameters");
+      for (const auto& argument : arguments) {
+        Assert(argument->data_type() == DataType::String || argument->data_type() == DataType::Null, "Concatenate takes only Strings and Nulls as arguments");
+      }
+      break;
+  }
+}
+
+std::shared_ptr<AbstractExpression> FunctionExpression::deep_copy() const {
+  return std::make_shared<FunctionExpression>(function_type, expressions_copy(arguments));
+}
+
+std::string FunctionExpression::as_column_name() const {
+  std::stringstream stream;
+
+  stream << function_type_to_string.left.at(function_type) << "(";
+  for (auto argument_idx = size_t{0}; argument_idx < arguments.size(); ++argument_idx) {
+    stream << arguments[argument_idx]->as_column_name();
+    if (argument_idx + 1 < arguments.size()) stream << ", ";
+  }
+  stream << ")";
+  return stream.str();
+}
+
+DataType FunctionExpression::data_type() const {
+  switch (function_type) {
+    case FunctionType::Substring: return DataType::String;
```

Comment:
yes, and because it's always a good idea to use a switch for enum classes because the compiler will error/warn you about enumeration member not listed.